### ISSUE TITLE
Segfault when deleting more than 1 item

### ISSUE THREAD

**Issue by sloriot** [CGAL/cgal#236]
In the polyhedron demo, If you select more than one item and remove them, the demo will crash with a segmentation fault.

cc @maxGimeno

**Comment by lrineau** [comment]
The end of the backtrace is:
```
#1  0x00007ffff49855c9 in QMetaObject::cast(QObject*) const (this=this@entry=0x7fff4b339a00 <Scene_hole_visualizer::staticMetaObject>, obj=0xf961b0) at kernel/qmetaobject.cpp:331
#2  0x00007fff4b0af440 in Polyhedron_demo_hole_filling_plugin::item_about_to_be_destroyed(Scene_item*) (object=<optimized out>) at /usr/include/qt5/QtCore/qobject.h:520
#3  0x00007fff4b0af440 in Polyhedron_demo_hole_filling_plugin::item_about_to_be_destroyed(Scene_item*) (poly_item=0x10aaee0, this=0xc9d420) at /home/lrineau/Git/cgal/Polyhedron/demo/Polyhedron/Polyhedron_demo_hole_filling_plugin.cpp:312
#4  0x00007fff4b0af440 in Polyhedron_demo_hole_filling_plugin::item_about_to_be_destroyed(Scene_item*) (this=0xc9d420, scene_item=<optimized out>) at /home/lrineau/Git/cgal/Polyhedron/demo/Polyhedron/Polyhedron_demo_hole_filling_plugin.cpp:429
#5  0x00007fff4b0b10d5 in Polyhedron_demo_hole_filling_plugin::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) (_o=0xc9d420, _c=<optimized out>, _id=<optimized out>, _a=<optimized out>) at /home/lrineau/Git/cgal/build/demo/Polyhedron/Polyhedron_demo_hole_filling_plugin.moc:209
#6  0x00007ffff49ae0ee in QMetaObject::activate(QObject*, int, int, void**) (sender=sender@entry=0xbf1b20, signalOffset=<optimized out>, local_signal_index=local_signal_index@entry=3, argv=argv@entry=0x7fffffffc2c0) at kernel/qobject.cpp:3717
#7  0x00007ffff49ae767 in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) (sender=sender@entry=0xbf1b20, m=m@entry=0x7ffff52a5a00 <Scene::staticMetaObject>, local_signal_index=local_signal_index@entry=3, argv=argv@entry=0x7fffffffc2c0) at kernel/qobject.cpp:3582
#8  0x00007ffff509abca in Scene::itemAboutToBeDestroyed(Scene_item*) (this=this@entry=0xbf1b20, _t1=0x10aaee0) at /home/lrineau/Git/cgal/build/demo/Polyhedron/moc_Scene.cpp:292
#9  0x00007ffff508b7c7 in Scene::erase(QList<int>) (this=0xbf1b20, indices=...) at /home/lrineau/Git/cgal/Polyhedron/demo/Polyhedron/Scene.cpp:119
```

The code of `Polyhedron_demo_hole_filling_plugin::get_hole_visualizer` is:
```C++
  Scene_hole_visualizer* get_hole_visualizer(Scene_polyhedron_item* poly_item) {
    // did not use a map to assoc Scene_polyhedron_item with Scene_hole_visualizer to prevent crowded code
    for(Scene_interface::Item_id i = 0, end = scene->numberOfEntries(); i < end; ++i) {
      Scene_hole_visualizer* hole_visualizer = qobject_cast<Scene_hole_visualizer*>(scene->item(i));
      if(hole_visualizer && hole_visualizer->poly_item == poly_item) 
      { return hole_visualizer; }
    }
    return NULL;
  }
```

During the segfault, the `i` of the loop is `0`. If I go up in the stack, at `Scene::erase`, I see:
```C++
int
Scene::erase(QList<int> indices)
{
    QList<Scene_item*> to_be_removed;

    int max_index = -1;
    Q_FOREACH(int index, indices) {
        if(index < 0 || index >= m_entries.size())
            continue;
        max_index = (std::max)(max_index, index);
        Scene_item* item = m_entries[index];
        to_be_removed.push_back(item);
        Q_EMIT itemAboutToBeDestroyed(item);   // WE ARE HERE
        delete item;
    }



  Q_FOREACH(Scene_item* item, to_be_removed) {
    m_entries.removeAll(item);
  }

  selected_item = -1;
  QAbstractListModel::beginResetModel();
  Q_EMIT updated();
  QAbstractListModel::endResetModel();
 
  int index = max_index + 1 - indices.size();
  if(index >= m_entries.size()) {
    index = m_entries.size() - 1;
  }
  if(index >= 0)
    return index;
  if(!m_entries.isEmpty())
    return 0;
  return -1;
```

The segfault is where I have added `WE ARE HERE`. In that frame, the `index` of the loop is `1`. That means that the item 0 has already been deleted. ... Thus the segfault.

The function `Polyhedron_demo_hole_filling_plugin::get_hole_visualizer` should be completely rewrote. The pointer to the hole visualizer should be stored as a dynamic property in the item itself:
```C++
  poly_item->setProperty("hole_visualizer", QVariant::fromValue((QObject*)hole_visualizer);
  // ...
  QObject* hole_visualizer_base = poly_item.property("hole_visualizer").value<QObject*>();
  Scene_hole_visualizer* hole_visualizer = 0;
  if(hole_visualizer_base && hole_visualizer = qobject_cast<Scene_hole_visualizer*>(hole_visualizer_base)) {
    // do something with hole_visulizer
  }
```

**Comment by sloriot** [comment]
I found a simpler fix: 5db1450

**Comment by lrineau** [comment]
5db1450981a2b5aef27ab4b1734fec1494ffa484 is already merged in cgal/master
